specs/psycopg_analysis.md

```diff
@@ -0,0 +1,176 @@
+
+## Discussion
+
+Explicit type declarations provide a clear and unambiguous way to describe the types of the symbols that comprise a library's public interface. When a type declaration is omitted, type checkers can fall back on a variety of type inference techniques to infer the intent of the library author, but these techniques are not standardized across type checkers. Depending on the circumstances, inferred type behaviors can vary greatly from one tool to the next. Even within a single type checker, type inference behaviors can change over time as improvements are made. This "type inference ambiguity" can lead to a bad experience for consumers of the library.
+
+Currently, Pyright assumes that any package marked as "py.typed" will provide unambiguous type declarations for all of its public symbols. Any symbols that are left unannotated (with a few specific exceptions spelled out in the documentation) are interpreted by Pyright as `Unknown`. The intent behind this approach is to increase the visibility of the missing type annotations so library authors can take action and eliminate the holes.
+
+While this approach has proven successful at raising the visibility with library authors, it has two significant downsides:
+
+1. It effectively "punishes" users of Pyright and Pylance who consume "py.typed" libraries when type annotations are missing. When referencing these unannotated symbols, they receive no completion suggestions, and they receive difficult-to-correct errors when referencing these symbols in "strict" static type checking mode. Users have little or no recourse other than to report the problem to the library maintainer and hope that it is addressed in the next package update.
+2. Some library maintainers are reluctant to add explicit type annotations in cases where they feel there is low risk of type inference ambiguity. This leaves them in an awkward situation when consumers of their library express dissatisfaction.
+
+It is perhaps time to consider an alternative approach, one that better balances the interests of users and library authors.
+
+## Recommendations
+
+I'm recommending the following changes:
+
+### Pyright's treatment of unannotated symbols in "py.typed" libraries
+
+Today, Pyright treats unannotated symbols in a "py.typed" library as "Unknown" if type checking is enabled. When typeCheckingMode is "off", it falls back to its inference logic. This effectively "punishes" users who want to use Pyright and Pylance for static type checking. I recommend that we change this behavior and always fall back on type inference logic when annotations are not present. This change will unfortunately reduce the visibility of missing annotations, so I worry that it will slow efforts to improve type completeness and consistency across the Python ecosystem, but I think it represents a pragmatic tradeoff.
```

Comment:
Will this continue to be controlled by, `useLibraryCodeForTypes`? I have that set to False as inferred types for tensorflow produce a good number of false positives that they've caused a fair amount of annoyance.

Reply:
This document is specifically talking about "py.typed" libraries. The `useLibraryCodeForTypes` setting does not apply to "py.typed" libraries. Since tensorflow is not currently "py.typed", `useLibraryCodeForTypes` would still apply.